### PR TITLE
Add Infoblox API support

### PR DESCRIPTION
**Pull request by taishan69**
Written in concert with one of my colleagues, Elijah Tenai, this script adds Infoblox API support to acme.sh.  Please review at your leisure.  Thank you for such a great script for Let's Encrypt support!

### CODE REVIEW

File: dns_infoblox.sh [hunk 1]
```diff
@@ -0,0 +1,78 @@
+
```
Comment (Neilpang) [inline dns_infoblox.sh]: please remove the first empty line.

**Comment by Neilpang** [comment]
Please fix the errors:

https://travis-ci.org/Neilpang/acme.sh/builds/203847213

**Comment by taishan69** [comment]
Truthfully, I don't know exactly what needs to be fixed.  I'm actively running it under Linux (RHEL7) in our prod environment with no errors or warnings, and the error is somewhat cryptic (what environment variable is it complaining about not having?).  I should probably mention that I'm not a programmer by trade, nor is my partner for this project, so please pardon any gaffes.

**Comment by Neilpang** [comment]
do not use `[[  ]]`, it's bash only.   please use `[ ]` instead.

**Comment by taishan69** [comment]
Sorry that took so long to get back to - I've removed all [[ ]] down to [ ].  Thanks @Neilpang for the additional guidance on it.

**Comment by Neilpang** [comment]
@taishan69   can you please try again, is it working ?

**Comment by taishan69** [comment]
@Neilpang sorry it did not work at all the way I had it before; I read up on regex conversion from bash/sh and have modified and tested it; recommitted with tested working changes.

File: dns_infoblox.sh [hunk 2]
```diff
@@ -24,54 +29,58 @@ dns_infoblox_add() {
 
 result=`curl -k -u $Infoblox_Creds -X POST $baseurlnObject`
 
-if [ $result =~ record:txt/.*:.*/default ]; then
+if echo "$result" | grep -Eq 'record:txt/.*:.*/default'; then
   echo "Successfully created the txt record"
```
Comment (Neilpang) [inline dns_infoblox.sh]: do not use `echo` to output messages.
use `_info()` function instead.

**Comment by taishan69** [comment]
@Neilpang changes committed; tested working

File: dns_infoblox.sh [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    return 1
+  fi
+
+  #save the login info to the account conf file.
+  _saveaccountconf Infoblox_Creds "$Infoblox_Creds"
+  _saveaccountconf Infoblox_Server "$Infoblox_Server"
+
+result=`curl -k -u $Infoblox_Creds -X POST $baseurlnObject`
+
+if _info "$result" | grep -Eq 'record:txt/.*:.*/default'; then
```
Comment (Neilpang) [inline dns_infoblox.sh]: do not use `grep -E`,  it's not supported on solaris.
Please use `egrep` instead.

**Comment by taishan69** [comment]
@Neilpang changes tested and committed

File: dns_infoblox.sh [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+  _saveaccountconf Infoblox_Server "$Infoblox_Server"
+
+result=`curl -k -u $Infoblox_Creds -X POST $baseurlnObject`
+
+if _info "$result" | egrep 'record:txt/.*:.*/default'; then
+  _info "Successfully created the txt record"
+  return 0
+else
+  _info "Error encountered during record addition"
+  _info $result
```
Comment (Neilpang) [inline dns_infoblox.sh]: use double quotation marks :

```
_info  "$result"
```

File: dns_infoblox.sh [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+  _debug fulldomain "$fulldomain"
+
+  _debug txtvalue "$txtvalue"
+
+ # Does the record exist?
+
+baseurlnObject="https://$Infoblox_Server/wapi/v2.2.2/record:txt?name=$fulldomain&text=$txtvalue&_return_type=xml-pretty"
+
+_info $baseurlnObject
```
Comment (Neilpang) [inline dns_infoblox.sh]: quotation marks

File: dns_infoblox.sh [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+               return 0
+        else
+            _info "Error occurred during txt record delete"
+            _info  $rmResult
+            _err $rmResult
+            return 1
+        fi
+else
+  _info "Record to delete didn't match an existing record"
+  _info $result
```
Comment (Neilpang) [inline dns_infoblox.sh]: quotation marks

**Comment by taishan69** [comment]
@Neilpang quotations added; tested and committed

**Comment by Neilpang** [comment]
Please fix the format errors:  https://travis-ci.org/Neilpang/acme.sh/jobs/211400574

**Comment by taishan69** [comment]
@Neilpang done; tested working

**Comment by taishan69** [comment]
@Neilpang looks like it's all checking out in TravisCI and ready for final review/merge

File: dns_infoblox.sh [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    _err "You didn't specify the credentials or server yet (Infoblox_Creds and Infoblox_Server)."
+    _err "Please set them via EXPORT ([username:password] and [ip or hostname]) and try again."
+    return 1
+  fi
+
+  #save the login info to the account conf file.
+  _saveaccountconf Infoblox_Creds "$Infoblox_Creds"
+  _saveaccountconf Infoblox_Server "$Infoblox_Server"
+
+  result=$(curl -k -u $Infoblox_Creds -X POST $baseurlnObject)
```
Comment (Neilpang) [inline dns_infoblox.sh]: do not use `curl` or `wget` in your code.
Please use `_post()` function to do post request.

File: dns_infoblox.sh [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+  _debug txtvalue "$txtvalue"
+
+  # Does the record exist?
+
+  baseurlnObject="https://$Infoblox_Server/wapi/v2.2.2/record:txt?name=$fulldomain&text=$txtvalue&_return_type=xml-pretty"
+
+  _info "$baseurlnObject"
+
+  result=$(curl -k -u $Infoblox_Creds -X GET $baseurlnObject)
```
Comment (Neilpang) [inline dns_infoblox.sh]: Use `_get()` function

File: dns_infoblox.sh [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+
+  if _info "$result" | egrep 'record:txt/.*:.*/default'; then
```
Comment (Neilpang) [inline dns_infoblox.sh]: do not use `_info()` function here.
If `_info()` is only used to output message to the user.
you can use `echo` here.

File: dns_infoblox.sh [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+    # Extract object ref
+    objRef=$(grep -Po 'record:txt/.*:.*/default' <<<$result)
+    objRmUrl="https://$Infoblox_Server/wapi/v2.2.2/$objRef"
+    rmResult=$(curl -k -u $Infoblox_Creds -X DELETE $objRmUrl)
+    # Check if rm succeeded
+    if _info "$rmResult" | egrep 'record:txt/.*:.*/default'; then
```
Comment (Neilpang) [inline dns_infoblox.sh]: use `echo`

File: dns_infoblox.sh [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+  _info "$baseurlnObject"
+
+  result=$(curl -k -u $Infoblox_Creds -X GET $baseurlnObject)
+
+  if _info "$result" | egrep 'record:txt/.*:.*/default'; then
+    # Extract object ref
+    objRef=$(grep -Po 'record:txt/.*:.*/default' <<<$result)
+    objRmUrl="https://$Infoblox_Server/wapi/v2.2.2/$objRef"
+    rmResult=$(curl -k -u $Infoblox_Creds -X DELETE $objRmUrl)
```
Comment (Neilpang) [inline dns_infoblox.sh]: Use `_post()` function, and pass the `DELETE` method to it.

File: dns_infoblox.sh [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+
+  baseurlnObject="https://$Infoblox_Server/wapi/v2.2.2/record:txt?name=$fulldomain&text=$txtvalue&_return_type=xml-pretty"
+
+  _info "$baseurlnObject"
+
+  result=$(curl -k -u $Infoblox_Creds -X GET $baseurlnObject)
+
+  if _info "$result" | egrep 'record:txt/.*:.*/default'; then
+    # Extract object ref
+    objRef=$(grep -Po 'record:txt/.*:.*/default' <<<$result)
```
Comment (Neilpang) [inline dns_infoblox.sh]: do not use `grep -o`,  please use `_egrep_o()` function.

**Comment by taishan69** [comment]
@Neilpang not to sound like a broken record/snippy, but...

**I'm not a programmer**
**You have write access**

Unless I have clear, concise examples for these functions, I cannot change/add them.  It's working fine in my environment, so rather than spending 16 hours trying to reverse engineer all this cold not knowing the peculiarities of every single shell/language on earth, I'd much rather you just add the edits you're asking for (since you seem to know exactly what you want to see) and just ask me to test it.  Would go much faster.

**Comment by taishan69** [comment]
@Neilpang committed with additional comments and code changes

**Comment by taishan69** [comment]
@Neilpang also about my (deleted) comment Friday; when you want something done, you have to communicate clearly.  Your vague references did not clearly elucidate that you created your own **custom** functions in your source code.  Your comments are blatantly assuming the other person you are interacting with is a programmer, which I clearly stated twice is not the case here.

This comment, for instance...
> Use _get() function

Should have been...
> I have written a custom function _get() in my source code (please see https://github.com/Neilpang/acme.sh/blob/master/acme.sh for variables passed and syntax); please use this instead of curl

Would have saved over an hour of trawling the internet uselessly for information regarding a function that actually isn't a global function of **sh**, but rather a custom function within **acme.sh**.  I may be capable of programming, but making it harder for me to do it is only naturally going to cause additional friction when asking me to spend my time changing code to benefit your project.

**Comment by Neilpang** [comment]
@taishan69  Sorry for the confusion.

Yes,  I'm assuming anyone sending PR on Github is a programmer. It's reasonable.

Yes, I knew you are not professional programmer, you don't need to claim twice. 

And I'm assuming anyone who send PR to acme.sh, have already read the source code of acme.sh.  If one has never read acme.sh, I would doubt he's ready to contribute.

If you have any problem, like: what  "_get()" or "_post()" function is,  the easiest way is to reply on this issue to ask me, I would answer you.

**Comment by taishan69** [comment]
@Neilpang final review

File: dnsapi/dns_infoblox [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+
+  ## Save the credentials to the account file
+  _saveaccountconf Infoblox_Creds "$Infoblox_Creds"
+  _saveaccountconf Infoblox_Server "$Infoblox_Server"
+
+  ## Base64 encode the credentials
+  Infoblox_CredsEncoded=$(printf "$Infoblox_Creds" | _base64)
+
+  ## Construct the HTTP Authorization header
+  export _H2="Authorization: Basic $Infoblox_CredsEncoded"
```
Comment (Neilpang) [inline dnsapi/dns_infoblox]: please  use `_H1`, 
do not use `_H2` if `_H1` is not used.

File: dnsapi/dns_infoblox [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+
+  _info "Using Infoblox API"
+  _debug fulldomain "$fulldomain"
+  _debug txtvalue "$txtvalue"
+
+  ## Base64 encode the credentials
+  Infoblox_CredsEncoded=$(printf "$Infoblox_Creds" | _base64)
+
+  ## Construct the HTTP Authorization header
+  export _H2="Authorization: Basic $Infoblox_CredsEncoded"
```
Comment (Neilpang) [inline dnsapi/dns_infoblox]: please  use `_H1`, 
do not use `_H2` if `_H1` is not used.

File: dnsapi/dns_infoblox [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+
+  ## Does the record exist?  Let's check.
+  baseurlnObject="https://$Infoblox_Server/wapi/v2.2.2/record:txt?name=$fulldomain&text=$txtvalue&_return_type=xml-pretty"
+  result=$(_get "$baseurlnObject")
+
+  ## Let's see if we get something intelligible back from the grid
+  if echo "$result" | egrep 'record:txt/.*:.*/default'; then
+    ## Extract the object reference
+    objRef=$(egrep -o 'record:txt/.*:.*/default' <<<$result)
```
Comment (Neilpang) [inline dnsapi/dns_infoblox]: Do not use `egrep -o`,  we have a function `_egrep_o()`,  please use that instead.

Comment (Neilpang) [inline dnsapi/dns_infoblox]: `egrep -o` option is not supported on solaris.

Comment (Neilpang) [inline dnsapi/dns_infoblox]:
```
echo "$response" | _egrep_o  'record:txt/.*:.*/default'
```

File: dnsapi/dns_infoblox [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+    objRmUrl="https://$Infoblox_Server/wapi/v2.2.2/$objRef"
+    ## Delete them! All the stale records!
+    rmResult=$(_post "" "$objRmUrl" "" "DELETE")
+    ## Let's see if that worked
+    if echo "$rmResult" | egrep 'record:txt/.*:.*/default'; then
+      _info "Successfully deleted $objRef"
+      return 0
+    else
+      _info "Error occurred during txt record delete"
+      _info "$rmResult"
```
Comment (Neilpang) [inline dnsapi/dns_infoblox]: remove this `_info`,  just `_err` is enough

File: dnsapi/dns_infoblox [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+      _err "$rmResult"
+      return 1
+    fi
+  else
+    _info "Record to delete didn't match an existing record"
+    _info "$result"
```
Comment (Neilpang) [inline dnsapi/dns_infoblox]: remove this `_info "$result"`,  just `_err` is enough

File: dnsapi/dns_infoblox [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+  ## Add the challenge record to the Infoblox grid member
+  result=$(_post "" "$baseurlnObject" "" "POST")
+
+  ## Let's see if we get something intelligible back from the unit
+  if echo "$result" | egrep 'record:txt/.*:.*/default'; then
+    _info "Successfully created the txt record"
+    return 0
+  else
+    _info "Error encountered during record addition"
+    _info "$result"
```
Comment (Neilpang) [inline dnsapi/dns_infoblox]: remove this `_info`,  just `_err` is enough

**Comment by taishan69** [comment]
@Neilpang done

File: dnsapi/dns_infoblox [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+    objRef=$(_egrep_o 'record:txt/.*:.*/default' <<<$result)
+    objRmUrl="https://$Infoblox_Server/wapi/v2.2.2/$objRef"
+    ## Delete them! All the stale records!
+    rmResult=$(_post "" "$objRmUrl" "" "DELETE")
+    ## Let's see if that worked
+    if echo "$rmResult" | egrep 'record:txt/.*:.*/default'; then
+      _info "Successfully deleted $objRef"
+      return 0
+    else
+      _info "Error occurred during txt record delete"
```
Comment (Neilpang) [inline dnsapi/dns_infoblox]: use `_err` here instead of `_info`

File: dnsapi/dns_infoblox [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+      _err "$rmResult"
+      return 1
+    fi
+  else
+    _info "Record to delete didn't match an existing record"
```
Comment (Neilpang) [inline dnsapi/dns_infoblox]: use `_err` here instead of `_info`

**Comment by Neilpang** [comment]
Please update the 2 readme.md files too:
https://github.com/Neilpang/acme.sh/blob/master/README.md
https://github.com/Neilpang/acme.sh/blob/master/dnsapi/README.md

**Comment by taishan69** [comment]
@Neilpang done

**Comment by taishan69** [comment]
@Neilpang how do I do that? I'm googling all over the place but there doesn't seem to be anywhere in the web client to do this.

**Comment by taishan69** [comment]
@Neilpang See the new PR 742

**Pull request closed without merging**